examples/chat-example/src/main.rs

```diff
@@ -44,58 +41,54 @@ struct MyBehaviour {
 
 #[async_std::main]
 async fn main() -> Result<(), Box<dyn Error>> {
-    // Create a random PeerId
-    let id_keys = identity::Keypair::generate_ed25519();
-    let local_peer_id = PeerId::from(id_keys.public());
-    println!("Local peer id: {local_peer_id}");
-
-    // Set up an encrypted DNS-enabled TCP Transport over the yamux protocol.
-    let tcp_transport = tcp::async_io::Transport::new(tcp::Config::default().nodelay(true))
-        .upgrade(upgrade::Version::V1Lazy)
-        .authenticate(noise::Config::new(&id_keys).expect("signing libp2p-noise static keypair"))
-        .multiplex(yamux::Config::default())
-        .timeout(std::time::Duration::from_secs(20))
-        .boxed();
-    let quic_transport = quic::async_std::Transport::new(quic::Config::new(&id_keys));
-    let transport = OrTransport::new(quic_transport, tcp_transport)
-        .map(|either_output, _| match either_output {
-            Either::Left((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
-            Either::Right((peer_id, muxer)) => (peer_id, StreamMuxerBox::new(muxer)),
+    let mut swarm = libp2p::builder::SwarmBuilder::new()
+        .with_new_identity()
+        .with_async_std()
+        .with_tcp()
+        .without_tls()
+        .with_noise()
+        .with_other_transport(|key| {
+            quic::async_std::Transport::new(quic::Config::new(&key))
+                .map(|(peer_id, muxer), _| (peer_id, StreamMuxerBox::new(muxer)))
         })
-        .boxed();
+        .without_any_other_transports()
+        .with_behaviour(|key| {
```

Comment:
I think `with_behaviour` will need to be fallible?

Ideally though, we can design it such that the common case where it is not fallible doesn't "pay" for that by having to `Ok` wrap everything.

Reply:
I played around with making `with_behaviour` fallible.

A type parameter `E` is an option:

```
    pub fn with_behaviour<B, E>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, E>,
    ) -> Result<Builder<P, B>, E> {
```

Though even in the case where `constructor` explicitly returns an error, rustc is not able to infer the error type.

```
error[E0282]: type annotations needed
  --> examples/chat-example/src/main.rs:78:13
   |
78 |             Ok(MyBehaviour { gossipsub, mdns })
   |             ^^ cannot infer type of the type parameter `E` declared on the enum `Result`
   |
help: consider specifying the generic arguments
   |
78 |             Ok::<MyBehaviour, E>(MyBehaviour { gossipsub, mdns })
   |               ++++++++++++++++++

```

Unfortunately one can not provide a defautl value for the `E` type parameter.

```
error: defaults for type parameters are only allowed in `struct`, `enum`, `type`, or `trait` definitions
   --> libp2p/src/builder.rs:826:30
    |
826 |     pub fn with_behaviour<B, E = void::Void>(
    |                              ^^^^^^^^^^^^^^
    |
    = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
    = note: for more information, see issue #36887 <https://github.com/rust-lang/rust/issues/36887>
    = note: `#[deny(invalid_type_param_default)]` on by default
```

What does work is using `Box<dyn std::error::Error>`.

``` rust
    pub fn with_behaviour<B>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, Box<dyn std::error::Error>>,
    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
```

@thomaseizinger let me know if you are fine with `Box<dyn std::error::Error>`.

---

> Ideally though, we can design it such that the common case where it is not fallible doesn't "pay" for that by having to `Ok` wrap everything.

I see the motivation. I don't see how to do it.

I tried type overloading:

``` rust
    pub fn with_behaviour<B>(
        self,
        mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> B,
    ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
        self.with_behaviour(|key| Ok(constructor(key)))
    }
```

but that is not legal:

```
error[E0592]: duplicate definitions with name `with_behaviour`
   --> libp2p/src/builder.rs:832:9
    |
826 | /     pub fn with_behaviour<B>(
827 | |         self,
828 | |         mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> B,
829 | |     ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
    | |__________________________________________________________- other definition for `with_behaviour`
...
832 | /         pub fn with_behaviour<B>(
833 | |             self,
834 | |             mut constructor: impl FnMut(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> Result<B, Box<dyn std::error::Erro...
835 | |         ) -> Result<Builder<P, B>, Box<dyn std::error::Error>> {
    | |______________________________________________________________^ duplicate definitions for `with_behaviour`
```

Reply:
I pushed a commit that I think fixes this reasonably well.

Reply:
:rocket: thank you @thomaseizinger!